Support service-level node queries (WIP)

noelleleigh opened the pull request:
Fixes #2 

My strategy here is to strip the `query { node }` fields from each service AST, so the federator doesn't know about them and won't complain about the duplicates!

The only thing left is to implement the node queries and their resolvers in each service schema. I work in Graphene (Python), so I'm not familiar with how you would do that correctly. Can you give me some guidance?

noelleleigh committed:
add logic to remove existing node query

noelleleigh committed:
Update TODO

victorandree commented:
Hmm... Thanks! This one is a little bit less straightforward, I think.

- This does 2 visits over each `service.typeDefs`. I feel it should be possible to both clean up the AST _and_ collect node types in a single go?
- Strictly speaking, you want to filter the `node` field from the schema's root query type, which may not be called "Query". In that case, you'd have a `SchemaDefinition` which points this out. But it's parenthetical, probably.
- Your code isn't formatted like mine. I haven't configured [prettier](https://www.npmjs.com/package/prettier) for this project, but the code should be run through that with `{"singleQuote": true, "trailingComma": "all", "semi": true }`

To support `node` on each service, each service should just define a `node` resolver for the types it supports. This is basically part of their regular schema, so it's up to them to resolve it. They should probably throw if they get a Node ID of a type they don't support.

For example, `server-product.js` should

- Add a `node(id: ID!): Node` resolver to its schema definition
- Add a resolver which does something similar to the global Node resolver (https://github.com/victorandree/apollo-federation-relay/blob/master/node-gateway.js#L32):

  1. Decode the ID
  2. Validate that the service has the typename
  3. Load the requested entity

noelleleigh committed:
prettier

noelleleigh committed:
Add node resolvers to services

noelleleigh committed:
Merge branch 'master' into support-service-node-queries

noelleleigh committed:
prettier

noelleleigh committed:
do everything in one AST pass

noelleleigh commented:
Alright, here's some improvements

- Prettier applied to `node-gateway.js`
- All the AST operations are in a single `vist( { ObjectTypeDefinition } )` block. I added whitespace to group the two different steps visually.
- I tried to find a `SchemaDefinition` in the ASTs, but it didn't appear for whatever reason, even if I explicitly specified a `schema` in the SDL. So this PR is just gonna rely on the [Default Root Operation Type Names](https://spec.graphql.org/draft/#sec-Root-Operation-Types.Default-Root-Operation-Type-Names).
- Added resolvers for `node` in each service.

noelleleigh committed:
remove caveat from README

victorandree commented on node-gateway.js:
```diff
@@ -63,17 +60,32 @@ class NodeGateway extends ApolloGateway {
-          // We don't need any resolvers for these modules; they're just
-          // simple objects with a single `id` property.
-          modules.push({ typeDefs: toTypeDefs(name) });
-          seenNodeTypes.add(name);
```
You don't need an `else` after an `if` that you've returned in. I'd suggest inverting the second condition so the branches are on "equal footing". Something like this?

```typescript
service.typeDefs = visit(service.typeDefs, {
  ObjectTypeDefinition(node) {
    const name = node.name.value;

    // Remove existing `query { node }` from service to avoid collisions
    if (name === 'Query') {
      return visit(node, {
        FieldDefinition(node) {
          if (node.name.value === 'node') {
            return null;
          }
        },
      });
    }

    // Add any new Nodes from this service to the Node service's modules
    if (isNode(node) && !seenNodeTypes.has(name)) {
      // We don't need any resolvers for these modules; they're just
      // simple objects with a single `id` property.
      modules.push({ typeDefs: toTypeDefs(name) });
      seenNodeTypes.add(name);

      return;
    }
  },
});
```

noelleleigh replied:
I like it! Added 😄

victorandree reviewed (approved):
Looks nice! I have a small stylistic comment on the visit code, but it's no biggie. Thanks for your contributions :)

noelleleigh committed:
flip logic
